web_server.go

```diff
@@ -58,20 +60,55 @@ func newWebServer(
 		return nil, fmt.Errorf("failed reading default index.html page: %v", err)
 	}
 
-	fs := handleStatic([]byte(mungeIndexPage(baseURL, string(indexPage))), http.StripPrefix(baseURL, http.FileServer(http.Dir(webDir))), header)
+	nonce, err := generateNonce()
+	if err != nil {
+		return nil, fmt.Errorf("failed generating nonce: %v", err)
+	}
+
+	indexHTML, err := mungeIndexPage(baseURL, string(indexPage), nonce)
+	if err != nil {
+		return nil, fmt.Errorf("failed munging default index.html page: %v", err)
+	}
+
+	cspHeader := header.Get("Content-Security-Policy")
+	if cspHeader != "" {
+		if strings.Contains(cspHeader, "script-src") {
+			cspHeader = strings.ReplaceAll(cspHeader, "script-src", "script-src 'nonce-"+nonce+"'")
+		} else {
+			cspHeader = cspHeader + "; script-src 'nonce-" + nonce + "'"
+		}
+		if strings.Contains(cspHeader, "style-src") {
+			cspHeader = strings.ReplaceAll(cspHeader, "style-src", "style-src 'nonce-"+nonce+"'")
+		} else {
+			cspHeader = cspHeader + "; style-src 'nonce-" + nonce + "'"
+		}
+		header.Set("Content-Security-Policy", cspHeader)
+	}
+
+	fs := handleStatic([]byte(indexHTML), http.StripPrefix(baseURL, http.FileServer(http.Dir(webDir))), header)
 	if auth != nil {
 		fs = auth.wrap(fs)
 	}
 	return &WebServer{site, broker, fs, keychain, maxRequestSize, baseURL}, nil
 }
 
-func mungeIndexPage(baseURL, html string) string {
-	// HACK
-	// set base URL as a body tag attribute, to be used by the front-end for deducing hash-routing and websocket addresses.
-	html = strings.Replace(html, "<body", `<body data-base-url="`+baseURL+`"`, 1)
+func generateNonce() (string, error) {
+	// Generate a random nonce for Content Security Policy (CSP).
+	nonce := make([]byte, 32)
+	if _, err := rand.Read(nonce); err != nil {
+		return "", fmt.Errorf("failed reading default index.html page: %v", err)
+	}
+	return base64.StdEncoding.EncodeToString(nonce), nil
+}
+
+func mungeIndexPage(baseURL, html, nonce string) (string, error) {
+	// HACK: Set base URL as a body tag attribute, to be used by the front-end for deducing hash-routing and websocket addresses.
+	html = strings.Replace(html, "<body", `<body data-base-url="`+baseURL+`" data-nonce="`+nonce+`"`, 1)
 	// ./wave-static/a/b/c.d -> /base-url/wave-static/a/b/c.d
 	html = strings.ReplaceAll(html, `="./wave-static/`, `="`+baseURL+"wave-static/")
```

Comment:
Once the UI is built, there is a `<script type="module" crossorigin src="./wave-static/index-658bc908.js"></script>` added into the `<header>` of `/wave/ui/build/index.html` file, which does not contain `nonce` attribute. Isn't this a possible vulnerability?

```js
<!doctype html>
<html lang="en">

<head>
...
  <title>H2O Wave</title>
  <script type="module" crossorigin src="./wave-static/index-658bc908.js"></script>
...
</head>
```

Reply:
> Isn't this a possible vulnerability?

It isn't. Nonce is only needed for inline scipts since they are vulnerable to XSS. The one you show above is a regular 1st party script with a content that is already set.

Reply:
Makes sense. Thanks for clarification!